docs/strategy_reviewable.md

```diff
@@ -0,0 +1,255 @@
+
+1. **instrument** - produces a list of OPCODEs in order of execution (this is the simplest one, because we expect for a particular program, this list is always identical - EVM's determinism)
+2. **measure all** - produces a list of measurements per instruction, in order of execution, _for all instructions in the program_.
+3. **measure one** - produce a single measurement for a chosen instruction, i.e. "measure Nth instruction from start of program"
+4. **measure total** - produce a single overarching measurement for the entire program execution
+    - for a single instruction we can then have a **measure pro-rated** value, whereby we "split" the measurement total into contributions proportional to the count of each OPCODE instructions (**NOTE** this way of counting might be controversial and requires a special approach to the sample program set)
+    - for a single instruction we can then have a **measure inferred** value, whereby we deduct the _a priori_ known measurements of other instructions from the total. E.g. we have a value for `PUSH1` and for a `PUSH1, PUSH1, ADD` program we can infer a measurement of `ADD`.
+
+When we speak of measurements, we have the following values in mind (to be decided):
+1. **time** - wall clock time (the most universal and likely one)
+2. **cycles** - CPU cycles using TSC (`RDTSC`, `gotsc` etc.)
+
+### Individual measurement vs entire program execution measurement
+
+The main question regarding the measurement of (CPU-bound) computational cost of instructions is, whether it is the individual instruction execution to be measured or an entire program execution, consisting of multiple instructions.
+
+**Pros - individual instructions**
+
+[See here](https://htmlpreview.github.io/?https://github.com/imapp-pl/gas-cost-estimator/blob/master/src/analysis/exploration.nb.html) for preliminary results.
+
+Here we measure the computational cost of each iteration of the inner loop of the EVM interpreter.
+- measurement is more granular and allows us to analyze the measurements without worrying about statistical errors
+- in particular, it allows for granular discrimination between cost of OPCODEs located in various circumstances ("surroundings")
+
+**Cons**
+
+- need to be careful about timer precision and overhead (preliminary results try to account for it)
+
+**Pros - entire program measurement**
+
+[See for example here](See https://notes.ethereum.org/@chfast/benchmarking-evm-instructions)
+
+- impact of timer precision and overhead on the measurements is much less problematic
+- program execution more resembling the production environment, no tracing/debug calls
+
+**Cons**
+
+- couples the intrinsic effect of instruction circumstances with their statistical impact on the estimator outcome
+- must be more careful about program generation, to not have a dataset with adverse statistical properties (e.g. we'd need a very homogenous and random program generator, which is hard given the need to balance the stack)
```

Comment:
For long running programs where "full" stack balancing is needed instructions are measured in groups which are difficult to split. I.e. every instruction increasing stack height must be countered by an instruction reducing the stack height. E.g. it is easy to measure the time of `DUP1 + POP` and `PUSH1 + POP` and you can also get time different of `PUSH1` and `DUP1`. But I don't see a way to get the time of `POP` from these or any other instruction pairs configurations.